arroyo/processing/strategies/dead_letter_queue/policies/produce.py

```diff
@@ -0,0 +1,89 @@
+import base64
+import json
+import time
+from collections import deque
+from concurrent.futures import Future
+from typing import Deque, Optional
+
+from arroyo.backends.kafka.consumer import KafkaPayload, KafkaProducer
+from arroyo.processing.strategies.dead_letter_queue.policies.abstract import (
+    DeadLetterQueuePolicy,
+    InvalidMessages,
+    Serializable,
+)
+from arroyo.types import Message, Topic
+from arroyo.utils.metrics import get_metrics
+
+MAX_QUEUE_SIZE = 5000
+
+
+class ProduceInvalidMessagePolicy(DeadLetterQueuePolicy):
+    """
+    Produces given InvalidMessages to a dead letter topic.
+    """
+
+    def __init__(self, producer: KafkaProducer, dead_letter_topic: Topic) -> None:
+        self.__metrics = get_metrics()
+        self.__dead_letter_topic = dead_letter_topic
+        self.__producer = producer
+        self.__futures: Deque[Future[Message[KafkaPayload]]] = deque()
+
+    def handle_invalid_messages(self, e: InvalidMessages) -> None:
+        """
+        Produces a message to the given dead letter topic for each
+        invalid message in the form:
+
+        {
+            "topic": <original topic the bad message was produced to>,
+            "reason": <why the message(s) are bad>
+            "timestamp": <time at which exception was thrown>,
+            "message": <original bad message>
+        }
+        """
+        for message in e.messages:
+            payload = self._build_payload(e, message)
+            self._produce(payload)
+        self.__metrics.increment("dlq.produced_messages", len(e.messages))
+
+    def _build_payload(self, e: InvalidMessages, message: Serializable) -> KafkaPayload:
+        if isinstance(message, bytes):
+            message = base64.b64encode(message).decode("utf-8")
+        data = json.dumps(
+            {
+                "topic": e.topic,
+                "reason": e.reason,
+                "timestamp": e.timestamp,
+                "message": message,
+            }
+        ).encode("utf-8")
+        return KafkaPayload(key=None, value=data, headers=[])
+
+    def _produce(self, payload: KafkaPayload) -> None:
+        """
+        Prune done futures from queue if filled, then asynchronously
+        produce the message, adding the process (future) to the queue.
+        """
+        self._prune_done_futures()
+        self.__futures.append(
+            self.__producer.produce(
+                destination=self.__dead_letter_topic, payload=payload
+            )
+        )
+
+    def _prune_done_futures(self) -> None:
+        """
+        Filter futures deque, should iterate only once except
+        for rare edge case all processes are still running.
+        """
+        while len(self.__futures) >= MAX_QUEUE_SIZE:
+            self.__futures = deque(
+                [future for future in self.__futures if not future.done()]
+            )
+
```

Comment:
This is not the way to wait for a future to complete.
This way creates a CPU intensive tight loop that keeps taking the list of futures, putting it in a new deque and repeat. This uses a lot more CPU than you need starving everything else that needs CPU.

the appropriate approach is :
- if self.__futures < MAX_QUEUE_SIZE -> return, you are ok
- if we are higher than that: call `self.__futures[0].result()`.   This call waits on the first future to complete. Since it waits on a lock, the thread does not use CPU for this. 
- When `result()` returns check if the result is an error or not. If it is an error raise an exception otherwise remove the future from the deque.

Also you should never need to rebuild the deque.

See more details about futures here https://docs.python.org/3/library/asyncio-future.html#future-object

Reply:
Fixed.

Seems `result()` only returns if it worked and would raise an exception if something breaks, so I didn't manually error check or raise anything.